use-async-queue.ts

```diff
@@ -1,22 +1,34 @@
 import { useState, useRef, useCallback, useEffect } from 'react';
 import nextTick from 'next-tick';
 
-/**
- * Hook for managing a queue with limited concurrency.
- * @param {number} concurrency - Maximum number of tasks to process
- * concurrently. Default 8.
- * @param {Function} inflight - Callback when a task changes from pending to
- * being processed.
- * @param {Function} done - Callback when a task is done being processed.
- * @param {Function} drain - Callback when all tasks are done.
- * @returns {object} Returns an object that exposes `add` and `stats`.
- */
-export default function useAsyncQueue({
-  concurrency = 8,
-  done,
-  drain,
-  inflight,
-}) {
+interface QueueStats {
+  numPending: number;
+  numInFlight: number;
+  numDone: number;
+}
+
+interface QueueTaskResult {
+  id: any;
+  task(): Promise<any>;
```

Comment:
Doesn't matter for JS callers, but this could be further specified by having `QueueTask` take in a type parameter here.

That would allow you to enforce that the task and result line up.

Reply:
Having the task and result line up makes sense to me. I'll have to think about how to express it -- will be a good exercise to build my typing skills. We can discuss!